src/pip/_internal/utils/misc.py

```diff
@@ -852,17 +853,34 @@ def enum(*sequential, **named):
     return type('Enum', (), enums)
 
 
-def remove_auth_from_url(url):
-    # Return a copy of url with 'username:password@' removed.
-    # username/pass params are passed to subversion through flags
-    # and are not recognized in the url.
+def redact_password_from_url(url):
+    """Return a copy of the url by redacting the password in case it was
+    present. The password will be replaced with '****'.
+    """
+    purl = urllib_parse.urlsplit(url)
+
+    redacted_netloc = purl.netloc
+    if purl.password:
```

Comment:
It looks like this new version will leak the password if it's the empty string:
```
>>> a = urllib.parse.urlsplit("https://user:@domain.tld/svn")
>>> a.password
''
```
If you want to use the password attribute, you should probably be checking for non-None.